internal/xds/rbac/rbac_engine.go

```diff
@@ -45,38 +122,112 @@ func NewEngine(policy *v3rbacpb.RBAC) (*Engine, error) {
 		}
 		policies[name] = matcher
 	}
-	return &Engine{policies: policies}, nil
+	return &engine{
+		policies: policies,
+		action:   action,
+	}, nil
 }
 
-// RPCData wraps data pulled from an incoming RPC that the RBAC engine needs to
+type connectionKey struct{}
+
+func getConnection(ctx context.Context) net.Conn {
+	conn, _ := ctx.Value(connectionKey{}).(net.Conn)
+	return conn
+}
+
+// SetConnection adds the connection to the context to be able to get
+// information about the destination ip and port for an incoming RPC.
+func SetConnection(ctx context.Context, conn net.Conn) context.Context {
+	return context.WithValue(ctx, connectionKey{}, conn)
+}
+
+// Data represents the generic data about incoming RPC's that must be passed
+// into the RBAC Engine in order to try and find a matching policy or not. The
+// ctx passed in must have metadata, peerinfo (used for source ip/port and TLS
+// information) and connection (used for destination ip/port) embedded within
+// it.
+type Data struct {
+	Ctx        context.Context
+	MethodName string
+}
+
+// newRPCData takes a incoming context (should be a context representing state
+// needed for server RPC Call with headers, connection and peer info piped into
+// it) and the method name of the Service being called server side and populates
+// an rpcData struct ready to be passed to the RBAC Engine to find a matching
+// policy.
+func newRPCData(data Data) (*rpcData, error) {
+	// The caller should populate all of these fields (i.e. for empty headers,
+	// pipe an empty md into context).
+	md, ok := metadata.FromIncomingContext(data.Ctx)
+	if !ok {
+		return nil, errors.New("error retrieving metadata from incoming ctx")
+	}
+
+	pi, ok := peer.FromContext(data.Ctx)
+	if !ok {
+		return nil, errors.New("error retrieving peer info from incoming ctx")
+	}
+
+	// The connection is needed in order to find the destination address and
+	// port of the incoming RPC Call.
+	conn := getConnection(data.Ctx)
```

Comment:
I tested this/thought about it. If conn is nil, this will panic on the net.SplitHostPort() line. The thing is, in order for there to be an RPC which is being intercepted, it has to happen on an established connection. Thus, I think it is a reasonable assumption that the connection is required in the context for this.

Reply:
There is still the possibility that in some codepath, the call to `SetConnection()` is missed. I wouldn't *not* protect myself from a `panic`, when it is simple enough to do.

Reply:
Whoops, missed this one. Added a bool return value, and returned an error similar to Metadata and PeerInfo.